main.tf

```diff
@@ -34,6 +33,73 @@ resource "aws_s3_bucket" "default" {
       }
     }
   }
+
+  dynamic lifecycle_rule {
+    for_each = var.lifecycle_rule
+
+    content {
+      id                                     = lookup(lifecycle_rule.value, "id", null)
+      prefix                                 = lookup(lifecycle_rule.value, "prefix", null)
+      tags                                   = lookup(lifecycle_rule.value, "tags", null)
+      abort_incomplete_multipart_upload_days = lookup(lifecycle_rule.value, "abort_incomplete_multipart_upload_days", null)
+      enabled                                = lifecycle_rule.value.enabled
+
+      # Max 1 block - expiration
+      dynamic expiration {
+        for_each = length(keys(lookup(lifecycle_rule.value, "expiration", {}))) == 0 ? [] : [lookup(lifecycle_rule.value, "expiration", {})]
+
+        content {
+          date                         = lookup(expiration.value, "date", null)
+          days                         = lookup(expiration.value, "days", null)
+          expired_object_delete_marker = lookup(expiration.value, "expired_object_delete_marker", null)
+        }
+      }
+
+      # Several blocks - transition
+      dynamic transition {
+        for_each = lookup(lifecycle_rule.value, "transition", [])
+
+        content {
+          date          = lookup(transition.value, "date", null)
+          days          = lookup(transition.value, "days", null)
+          storage_class = transition.value.storage_class
+        }
+      }
+
+      # Max 1 block - noncurrent_version_expiration
+      dynamic noncurrent_version_expiration {
+        for_each = length(keys(lookup(lifecycle_rule.value, "noncurrent_version_expiration", {}))) == 0 ? [] : [lookup(lifecycle_rule.value, "noncurrent_version_expiration", {})]
+
+        content {
+          days = lookup(noncurrent_version_expiration.value, "days", null)
+        }
+      }
+
+      # Several blocks - noncurrent_version_transition
+      dynamic noncurrent_version_transition {
+        for_each = lookup(lifecycle_rule.value, "noncurrent_version_transition", [])
+
+        content {
+          days          = lookup(noncurrent_version_transition.value, "days", null)
+          storage_class = noncurrent_version_transition.value.storage_class
+        }
+      }
+    }
+  }
+
+  replication_configuration {
+    role = var.repl_iam_role_arn
+
+    rules {
+      id     = "ReplicateSFTP2Raw"
+      status = "Enabled"
+
+      destination {
+        bucket        = var.repl_dest_bucket
+        storage_class = var.repl_storage_class
+      }
+    }
+  }
```

Comment:
As discussed currently the `replication_configuration` block is required when it may not be required, so you need a way to optionally create rules when a variable is populated.  Perhaps a `replication_config` map?

The `rules` block should be dynamic like the `lifecycle` rule (and could be a list in the `replication_config` map)

Reply:
Should we maybe have a short discussion about your intended result (what do you want to add) and a solution on how to get there? The current implementation seems very loosely defined and so it super hard to understand how it should be used and can very easily be used wrongly. So I hope we can make better use of the type system that TF offers to make it more clear what the options are for users of this module.

Reply:
further discussing in updated branch